### ISSUE TITLE
'Binding to native code using dart:ffi' page issue

### ISSUE THREAD

**Issue by Timmmm** [flutter/website#5334]
Page URL: https://flutter.dev/docs/development/platform-integration/c-interop.html
Page source: https://github.com/flutter/website/tree/master/src/docs/development/platform-integration/c-interop.md

Description of issue: The introduction isn't very clear about static vs dynamic linking. It explains what they are - fine, I expect most people reading it already know - but then it just jumps into dynamic linking on Android and static linking on iOS with no explanation as to why.

I *presume* that dynamic linking is forbidden on iOS and static linking doesn't work on Android? Or something like that? I would be good to explain that.

**Comment by sfshaza2** [comment]
Thoughts, @mit-mit?

**Comment by mit-mit** [comment]
Yes, I agree that this page would benefit from some elaboration.

**Comment by dcharkes** [comment]
Similar issue https://github.com/flutter/flutter/issues/75537.

> I presume that dynamic linking is forbidden on iOS and static linking doesn't work on Android? Or something like that? I would be good to explain that.

We have not explored the non-documented ways. They might be possible. If they are, PRs are welcome to update the documentation.

We should update the page to reflect this. @sfshaza2 @mit-mit any suggestions for wording?

**Comment by sfshaza2** [comment]
@chinmaygarde, do you have any insight into what we could say about this distinction?

**Comment by chinmaygarde** [comment]
There is no magic in the way symbol resolution works for Dart FFI. The symbols need to be resolved from a library using `dlsym`. `dlsym` needs a library handle. This handle is obtained by using `dlopen` with a specified path to a dynamic library. If symbol is already present in the current global scope, `RTLD_DEFAULT` can be used instead.

In Dart code, `DynamicLibrary.open()` is used to obtain the handle from a path and `DyanmicLibrary.process()` for the `RTLD_DEFAULT` handle.

The specific guidance for iOS and Android differs because of the way the applications are packaged.

On iOS, there is an executable that needs to be built already and it is generally easier to link static libraries into this executable. Once in the executable, `DynamicLibrary.process()` will work fine. However, there is nothing preventing the user from resolving the symbols from a dynamic library. I created a sample project that demonstrates this https://github.com/chinmaygarde/flutter_dylinkios_sample. Notice that in the `Build Phases`, I explicitly specified that the library be linked. This way, the symbols from that library are available in the global scope and I can use `DynamicLibrary.process`. But, I didn't need to do that. I could have skipped that step and loaded the library myself at runtime using `DynamicLibrary.open()` with the path returned by [`[NSBundle exectuablePath]`](https://developer.apple.com/documentation/foundation/nsbundle/1409078-executablepath) for the framework bundle. It would have achieved the same thing.  Using static libraries on iOS avoids having to remember to link the library and make sure to embed it in the final bundle, or, figure out some way of finding where the dynamic library exists at runtime. But, as demonstrated, it is supported and works fine.
![Screen Shot 2021-03-15 at 12 15 58 PM](https://user-images.githubusercontent.com/44085/111208846-85193400-8588-11eb-87cc-df7b4ba3aaab.png)

Now, on Android, there no executable generated (the thing that contains the `main` entrypoint). So, we have to rely on dynamic libraries only on Android.

Hope this helps clarify things.

**Comment by sfshaza2** [comment]
Wow, @chinmaygarde, thanks for that detailed answer!

**Comment by Timmmm** [comment]
Ah that makes sense. Although isn't Flutter AoT compiled now on Android? Does that not give a binary you can statically link into? I guess maybe it's just too much of a pain to do, and it wouldn't work with hot reload which would kind of suck.

**Comment by chinmaygarde** [comment]
Right, it would be limited to specific modes. Also, the default Flutter workflow for Android doesn't require a native toolchain. So everyone would have to download the NDK or some other toolchain and have that versioned and updated. The assumption is that if you are writing native code, you already have a build system and toolchain in place for it that can generate a dynamic library.
